components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,291 @@
+    if (type == "address") {
+      if (!GetAddressFromData(data, &value, offset))
+        return false;
+    } else if (type == "uint256") {
+      if (!GetUint256HexFromData(data, &value, offset))
+        return false;
+    } else if (type == "bool") {
+      if (!GetBoolFromData(data, &value, offset))
+        return false;
+    } else if (type == "bytes") {
+      if (!GetBytesHexFromData(data, &value, offset))
+        return false;
+    } else if (type == "address[]") {
+      if (!GetAddressArrayFromData(data, &value, offset))
+        return false;
+    } else {
+      // For unknown/unsupported types, we only extract 32-bytes. In case of
+      // dynamic types, this value is a calldata reference.
+      value = data.substr(offset, 64);
+    }
+
+    // On encountering a dynamic type, we extract the reference to the start
+    // of the tail section of the calldata.
+    if ((type == "bytes" || type == "string" ||
+         type.substr(type.length() - 2) == "[]") &&
+        calldata_tail == 0) {
+      size_t pointer;
+      if (!GetSizeFromData(data, &pointer, offset))
+        return false;
+
+      calldata_tail = pointer * 2;
+    }
+
+    offset += 64;
+
+    if (tx_args)
+      tx_args->push_back(value);
+
+    if (tx_params)
+      tx_params->push_back(type);
+  }
+
+  // Extraneous calldata, unless in the tail section, is considered invalid.
+  if (offset != calldata_tail && data.substr(offset).length() > 0)
+    return false;
+
+  return true;
+}
+
+}  // namespace
+
+bool UniswapEncodedPathDecode(const std::string& encodedPath,
+                              std::vector<std::string>* path) {
+  // Parse a Uniswap-encoded path and return a vector of addresses representing
+  // each hop involved in the swap.
+  //
+  // Single-hop swap: Token1 → Token2
+  // Multi-hop swap: Token1 → Token2 → WETH → Token3
+  //
+  // Each encoded hop contains a 3-byte pool fee, which is associated with the
+  // address that follows. It is excluded from the result of this function.
+  //
+  // ┌──────────┬──────────┬──────────┬─────┐
+  // │ address  │ pool fee │ address  │     │
+  // │          │          │          │ ... │
+  // │ 20 bytes │ 3 bytes  │ 20 bytes │     │
+  // └──────────┴──────────┴──────────┴─────┘
+  std::string data = encodedPath.substr(2);
+
+  // Parse first hop address.
+  if (data.length() < 40)
+    return false;
+  path->push_back("0x" + data.substr(0, 40));
+  data = data.substr(40);
+
+  while (true) {
+    if (data.length() == 0)
+      break;
+
+    // Parse the pool fee, and ignore.
+    if (data.length() < 6)
+      return false;
+    data = data.substr(6);
+
+    // Parse next hop.
+    if (data.length() < 40)
+      return false;
+    path->push_back("0x" + data.substr(0, 40));
+    data = data.substr(40);
+  }
+
+  // Require a minimum of 2 addresses for a single-hop swap.
+  if (path->size() < 2)
+    return false;
+
+  return true;
+}
+
+bool ABIDecode(const std::vector<std::string> types,
```

Comment:
Where these strings come from? Is it ok to parse them in browser process?

Reply:
`types` is hardcoded in calls to `ABIDecode` (see `eth_data_parser.cc`), so no surprises there and we only support a select few ABI types.

`calldata` on the other hand can originate from these sources:
1. transactions created from Brave Wallet → ex, `ERC20Transfer`, `ERC20Approve`.
2. transactions created by [0x API](https://0x.org/docs/api) → this is for swaps on Brave Wallet. Just like the above case, we already know the format to expect for each supported [function selector](https://docs.soliditylang.org/en/v0.8.12/abi-spec.html#function-selector).

    It is possible that the API returns a bad calldata - one that either has invalid hex (already handled before this PR), or one that doesn't match the expected ABI.
3. transactions created by dapps → we don't try to parse calldata of such transactions.

> Is it ok to parse them in browser process?

What's the alternative, and how to implement it? Sorry, I'm not very familiar with the C++ best practices here.